### DISCUSSION TITLE
Concurrency system

### DISCUSSION THREAD

**Discussion by foliot** [koka-lang/koka#227] [General]
Hi all, forgive me if this has been answered before, but what is the plan for concurrency/parallelism in Koka? Lightweight threads of some sort? A synchronous programming model, or async/await? 

The effect handlers and memory management of Koka are really cool, and I think with good concurrency support, it could be an excellent server language.

**Comment by daanx** [comment]
Hi @foliot ; yes, koka will be great for asynchronous programming as it fits perfectly with algebraic effect handlers with respect to composition and cancelation. (I wrote about an async design before [pdf](https://www.microsoft.com/en-us/research/wp-content/uploads/2017/05/asynceffects-msr-tr-2017-21.pdf)). 

Currently, I am about to start integrating [libuv](https://libuv.org) into koka and the plan is to use those primitives to build a nice async api on top using effect handlers. Hopefully we can experiment with this soon :-)

**Comment by foliot** [comment]
Awesome, thanks for the response!

**Comment by divs1210** [comment]
Any updates on this?

**Comment by cognivore** [comment]
@divs1210 

- https://github.com/koka-lang/koka/milestone/1
- https://github.com/koka-lang/koka/activity?ref=dev-uv
